services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java

```diff
@@ -452,28 +454,68 @@ public ClientPolicyManager clientPolicy() {
         return clientPolicyManager;
     }
 
+    private static final Logger LOG = Logger.getLogger(DefaultKeycloakSession.class);
+
     @Override
     public void close() {
+        if (LOG.isDebugEnabled()) {
+            LOG.debugf("Closing %s%s%s", this,
+              getTransactionManager().isActive() ? " (transaction active" + (getTransactionManager().getRollbackOnly() ? ", ROLLBACK-ONLY" : "") + ")" : "",
+              StackUtil.getShortStackTrace());
+        }
+
         if (closed) {
-            throw new IllegalStateException("Illegal call to #close() on already closed KeycloakSession");
+            throw new IllegalStateException("Illegal call to #close() on already closed " + this);
+        }
+
+        RuntimeException re = closeTransactionManager();
+
+        try {
+            Consumer<? super Provider> safeClose = p -> {
+                try {
+                    p.close();
+                } catch (Exception e) {
+                    // Ignore exception
+                }
+            };
+            providers.values().forEach(safeClose);
+            closable.forEach(safeClose);
+            for (Entry<InvalidableObjectType, Set<Object>> me : invalidationMap.entrySet()) {
+                factory.invalidate(this, me.getKey(), me.getValue().toArray());
+            }
+        } finally {
+            this.closed = true;
+        }
+
+        if (re != null) {
+            throw re;
+        }
+    }
+
+    protected RuntimeException closeTransactionManager() {
+        if (! this.transactionManager.isActive()) {
+            return null;
         }
-        Consumer<? super Provider> safeClose = p -> {
-            try {
-                p.close();
-            } catch (Exception e) {
-                // Ignore exception
+
+        try {
+            if (this.transactionManager.getRollbackOnly()) {
+                this.transactionManager.rollback();
+            } else {
+                this.transactionManager.commit();
+            }
+        } catch (RuntimeException re) {
+            if (this.transactionManager.isActive()) {
+                this.transactionManager.rollback();
```

Comment:
Seems this can call rollback twice, if it fails above?

Reply:
Upon rollback, TM will turn into inactive mode. Thus this line will not be reached. But you are right that this applies to commit too, so this condition is redundant and will be removed.